cohortextractor/query_language.py

```diff
@@ -44,6 +51,14 @@ def __or__(self, other):
     def __len__(self):
         return len(self.children)
 
+    def __invert__(self):
+        self.negate()
+        return self
+
+    def negate(self):
+        """Negate the sense of the root comparator."""
+        self.negated = not self.negated
+
```

Comment:
I'm a bit worried by the mutation here. I think we'll make life a lot easier for ourselves in the long run if we insist that everything in the query DAG is immutable and all methods like this return a new object with different properties.

Reply:
:+1: I can make it do that

Reply:
Also, I did wonder about making it use `is_null`/`not_null` instead, but the `c` function seemed a bit closer to how they write it at the moment (in the long covid study there's one that does `sgss_positive OR primary_care_covid AND NOT hospital_covid`).  But it's easy enough to change later if people prefer that

Reply:
Right, yes, good point. I suppose the ideal would be if we could do the Pythonic thing and implicitly convert non-boolean values to booleans if they're involved in boolean conditions. But I think that would require us tracking the types of each Value, which I'm sure we'll want to do eventually but will be quite a lot of work to add now.

Reply:
I dithered a bit over how to make the new object in the invert method, since it can't just use the `_combine` one (it could be negating a lone comparator or one with children).  I've used `deepcopy` but it could also more explicitly do something like
```
obj = type(self)(
    connector=self.connector,
    negated=not self.negated,
    source=self.source,
    operator=self.operator,
    value=self.value,
)
for child in self.children:
    obj.add(child, self.connector)
```
Any particular opinions either way?

Reply:
I can't remember exactly how `deepcopy` works, but I'd worry that it would end up duplicating the entire query tree i.e. it will see `self.source` and start copying that. I think maybe just shallow `copy.copy()` would do the right thing here. But no harm in doing in the explicit way.

Reply:
I feel like it ought to be possible for negation to be a generic operation on Values, rather than something specific to Comparator. Ultimately, we just need SQLAlchemy to call it's `not_` operator  on whatever the argument is. I don't quite have a clear picture of that would all work though so it's a thing to ponder rather than a suggestion for what to do now.

Reply:
Actually, I think we could have an `__invert__` method on a Value too. In fact, we should, so we can do `~foo` and don't have to do `~c(foo)`.